src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,207 @@
+        }
+        size++;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException("Index: " + index + " is out of bounds");
+        }
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
     @Override
     public void addAll(List<T> list) {
+        checkPositionIndex(size);
+        Object[] arrayOfList = list.toArray();
+        int numNew = arrayOfList.length;
+        if (numNew != 0) {
+            Node<T> predecessor = last;
+            for (Object object : arrayOfList) {
+                Node<T> newNode = new Node<>(predecessor, (T) object, null);
+                predecessor.next = newNode;
+                predecessor = newNode;
+            }
+            last = predecessor;
+            size += numNew;
+        }
     }
+    /////////////////////////////////////////////////////////////////////////////
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> x = node(index);
+        T oldVal = x.item;
+        x.item = value;
+        return oldVal;
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException("Index: " + index + "is out of bounds");
+        }
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
     }
 
+    //////////////////////////////////////////////////////////////////////////////
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
+    T unlink(Node<T> x) {
```

Comment:
```suggestion
    private T unlink(Node<T> node) {
```
use better names, remember about access modifiers

Reply:
Upgraded